controllers/reconcilers/kserve_isvc_gateway_reconciler.go

```diff
@@ -0,0 +1,263 @@
+
+			return nil
+		}
+	}
+
+	// Create Desired resource
+	desiredResource, err := r.createDesiredResource(isvc)
+	if err != nil {
+		return err
+	}
+
+	// Get Existing resource
+	existingResource, err := r.getExistingResource(ctx)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Error(err, "Failed to find KServe local gateway in istio-system namespace")
+		}
+		return err
+	}
+
+	// Process Delta
+	if err = r.processDelta(ctx, log, desiredResource, existingResource); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *KserveGatewayReconciler) createDesiredResource(isvc *kservev1beta1.InferenceService) (*istioclientv1beta1.Gateway, error) {
+	hostname := fmt.Sprintf("%s.%s.%s", isvc.Name, isvc.Namespace, internalSerivceHostnameDomain)
+
+	desiredGateway := &istioclientv1beta1.Gateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "kserve-local-gateway",
+			Namespace: isvc.Namespace,
+		},
+		Spec: istiov1beta1.Gateway{
+			Servers: []*istiov1beta1.Server{
+				{
+					Hosts: []string{hostname},
+					Port: &istiov1beta1.Port{
+						Name:     isvc.Name,
+						Number:   8445,
+						Protocol: "HTTPS",
+					},
+					Tls: &istiov1beta1.ServerTLSSettings{
+						CredentialName: isvc.Name,
+						Mode:           istiov1beta1.ServerTLSSettings_SIMPLE,
+					},
+				},
+			},
+		},
+	}
+
+	return desiredGateway, nil
+}
+
+func (r *KserveGatewayReconciler) getExistingResource(ctx context.Context) (*istioclientv1beta1.Gateway, error) {
+	return r.gatewayHandler.Get(ctx, types.NamespacedName{Name: constants.KServeGatewayName, Namespace: constants.IstioNamespace})
+}
+
+func (r *KserveGatewayReconciler) Delete(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error {
+	log.V(1).Info(fmt.Sprintf("Deleting serving cert secret(%s) in the namespace(%s)", isvc.Name, isvc.Namespace))
+	if err := r.deleteServingCertSecretInIstioNamespace(ctx, isvc.Name); err != nil {
+		log.V(1).Error(err, "Failed to delete the copied serving cert secret in the namespace")
+		return err
+	}
+
+	log.V(1).Info(fmt.Sprintf("Deleting the Server(%s) from KServe local gateway in the istio-system namespace", isvc.Name))
+	if err := r.removeServerFromGateway(ctx, log, isvc.Name); err != nil {
+		log.V(1).Error(err, "Failed to remove the server from KServe local gateway in the istio-system namespace")
+		return err
+	}
+
+	return nil
+}
+
+func (r *KserveGatewayReconciler) Cleanup(ctx context.Context, log logr.Logger, isvcName string) error {
+	// NOOP - Resources should not be deleted until the kserve component is uninstalled.
+	return nil
+}
+
+func (r *KserveGatewayReconciler) processDelta(ctx context.Context, log logr.Logger, desiredGateway *istioclientv1beta1.Gateway, existingGateway *istioclientv1beta1.Gateway) (err error) {
+	comparator := comparators.GetGatewayComparator()
+	delta := r.deltaProcessor.ComputeDelta(comparator, desiredGateway, existingGateway)
+
+	if !delta.HasChanges() {
+		// Note: This code will not be executed.
```

Comment:
expected?
If user deletes or change it by accident, what happens?

Reply:
Right. If user deletes the KServe local gateway, it doesn't do anything. The Gateway is managed by opendatahub.

When opendatahub creates the KServe local gateway again, it also does not do anything with the IsAdded().

But if the Gateway is created again, it will update the gateway using IsUpdated()